omfhint.cpp

```diff
@@ -142,7 +149,78 @@ OMFHintFilter::configure(const ConfigCategory& config)
 				} else {
 					m_hints.insert(pair<string, string>(asset, escaped));
 				}
+				// Check if macro substituion is required
+				// At least one pair of '$' sign must be there to apply macro
+				if (std::count(escaped.begin(), escaped.end(), '$') > 1)
+					isMacro = true;
 			}
 		}
+		if (isMacro)
+			collectMacrosInfo(config.getValue("hints"));
+	}
+}
+
+/**
+ * Extract datapoint name for macro replacement
+ *
+ * @param hintsJson	OMFHints JSON
+ */
+void OMFHintFilter::collectMacrosInfo(std::string hintsJSON)
+{
+	std::regex pattern("\\$([^$]*)\\$");
+	std::smatch match;
+
+	// Find and store all matching macros
+	while (std::regex_search(hintsJSON, match, pattern))
+	{
+		m_macro_dpName.emplace_back(match[1]);
+		hintsJSON = match.suffix();
+	}
+}
+
+/**
+ * Replace Macros with datapoint values
+ *
+ * @param reading	Reading
+ * @param hintsJson	OMFHints JSON
+ */
+void OMFHintFilter::ReplaceMacros(Reading *reading, std::string &hintsJSON)
+{
+	// Replace Macros by datapoint value
+	for ( auto dpName : m_macro_dpName)
+	{
+		// In case of ASSET Macro, replace it by asset name instead of datapoint value
+		if (dpName == "ASSET")
+		{
+			StringReplaceAll(hintsJSON, "$ASSET$", reading->getAssetName());
+			continue;
+		}
+		Datapoint * datapoint = reading->getDatapoint(dpName);
+
+		if (datapoint)
+		{
+			// Check for datapoint type for string and numbers
+			DatapointValue::dataTagType dataType = datapoint->getData().getType();
+			if (
+				dataType != DatapointValue::dataTagType::T_STRING &&
+				dataType != DatapointValue::dataTagType::T_INTEGER &&
+				dataType != DatapointValue::dataTagType::T_FLOAT
+			)
+			{
+				Logger::getLogger()->warn("Only string and number are supported for macro");
+				continue;
+			}
+
+			string datapointValue = datapoint->getData().toString();
+
+			// Strip quotes from begining and end
+			char quote = '"';
+			if (datapointValue.front() == quote && datapointValue.back() == quote)
```

Comment:
Why are you stripping quotes? You can use the toStringValue() method that doesn't add the quotes in the first place for string type datapoints and use the C++ toString method for numeric values. It will be much quicker and avoid making lots of copies of strings.

Reply:
Fixed